### ISSUE TITLE
Access to schema properties for non TObject schemas

### ISSUE THREAD

**Issue by saxomoose** [sinclairzx81/typebox#954]
I would like to create a generic type which allows me to point to any attribute of an existing type. This reference should include type name, attribute name and attribute range.

```typescript
const AttributeReference = <T extends TObject>(type: T, attribute: string) => {
    const typeName = type.$id as string;
    const attributeRange = type.properties[attribute].$id as string
    return Type.Object({
        type: Type.Literal(typeName),
        attributeName: Type.Literal(attribute),
        attributeRange: Type.Literal(attributeRange)
    })
}
```
Above attempt expects a TObject which excludes `Intersect<T>` or `Union<T>` schemas.

**Comment by sinclairzx81** [comment]
@saxomoose Hi,

> I would like to create a generic type which allows me to point to any attribute of an existing type. This reference should include type name, attribute name and attribute range.

Unfortunately, I don't fully understand the question. Are you able to show me an example usage of the above type and what you expect the output type to be? Also, if you can try model this type in TypeScript using https://sinclairzx81.github.io/typebox-workbench that would help me get a bit more insight into the type you are trying to build.

Let me know

**Comment by saxomoose** [comment]
The context is a statement-based data model. My goal is to make `LegalStatement` about `Record` in my system.
```typescript
const LegalStatement = Type.Object({
    legalEvent: Type.Union([core.Identifier, core.JsonPointer]),
    record: Type.Union([core.Identifier, core.JsonPointer]),
    value: Type.Any(),
    created: core.DateTime,
    effectiveFrom: core.TemporalPosition,
    effectiveUntil: Type.Optional(core.TemporalPosition),
})
```
A `LegalStatement` is made about a `Record`, target a particular attribute and carry a value. 

An effective statement would be an intersect of the above with a generic `AttributeReference(type: TSchema, attribute: string)`.
Such a function would allow me to:
* ensure that statements only target existing attributes of the underlying model. 
* indicate the range of the attribute so that the value can be parsed adequately.

The attempt at a function definition in my initial comment works if the schema is a `TObject` but fails if the schema is a `TIntersect<T>` or `TUnion<T>` since they lack a `properties` property.

Basically looking for a way to access the top-level properties of any typebox schema.

**Comment by sinclairzx81** [comment]
@saxomoose Hiya, 

Still uncertain about the question, but you can use Type.Index to access top level types embedded in Objects, Unions, Intersections and Tuples. This type is designed inline with [Indexed Access Types](https://www.typescriptlang.org/docs/handbook/2/indexed-access-types.html) in TypeScript

[TypeScript Link Here](https://www.typescriptlang.org/play/?#code/JYWwDg9gTgLgBAbzgFQJ5gKYBo4GUYCGMwAxnAL5wBmUEIcA5AAIDOwAdiQDYHBQD0MdBgBGEAB4MAUFJIR2LeMjgBeFMIB0AeREArDCRgAKBFLjm4BAFzrMGgDLAYGKAS5GAjAEosZiyJs0O0dnV3cAJh8-cxJAzRCXNyMAZiiLOAATOOCnRPcAFjSLDGyMB1ywowBWIvMqUvLQpIA2KPIvGX5+OABJdgyMcTgAQRISDBYWWwmZOQV4YdVpjT6B8SNkHABtBgIGAF0OqS6RwISwgB4PAD5Z+UU4ACEloLLVwY3t3YYcBhEDo4nOYPR6BACq7GA8guW2Q5zcV2uODhFQR4Wu+1usnu8AAwi9NO91ptlgBpDCoLRUDZeQHdYF48GQ6Gw+FcRHItkXdGc1Hs5JIlBc-KClFNdlVUVc5oY65AA)

```typescript
import { Type, Static } from '@sinclair/typebox'

const T = Type.Object({
    a: Type.Literal(1),
    b: Type.Literal(2),
    c: Type.Literal(3),
    d: Type.Literal(4),
    e: Type.Literal(5),
    f: Type.Literal(6),
})

// Index Access Types

const A = Type.Index(T, ['a'])

// A: TLiteral<1>

const B = Type.Index(T, ['a', 'b'])

// const B: TUnion<[TLiteral<1>, TLiteral<2>]>

const C = Type.Index(T, Type.KeyOf(T))

// const C: TUnion<[TLiteral<1>, TLiteral<2>, TLiteral<3>, TLiteral<4>, TLiteral<5>, TLiteral<6>]>
```
With respect to the original `AttributeReference` type, the TypeBox API doesn't provide facilities for assigning and accessing interior type properties (like `$id`) to construct new types. Instead it prefers you work through the `Type.*` builder only as this is aligned to TypeScript and can generate correct inference and validation for you. 

This said, while you can still generate types using the approach used to construct `AttributeReference`, you will likely need to use Type.Unsafe or explicit type annotations to hint at the exact schematics and static types you want. TypeBox is very closely aligned to the TypeScript type programming model, so if you can construct the equivalent type using only TypeScript type syntax, there is a reasonable chance TypeBox will be able to generate the correct type for you.

[Example Type Level Programming](https://sinclairzx81.github.io/typebox-workbench/?share=Ly8gLS0tLS0tLS0tLS0tLS0tLS0tLS0tLS0tLS0tLS0tLS0tLS0tLS0tLS0tLS0tLS0tLS0tLQovLyBUeXBlQm94IE1hcHBlZCBUeXBlcyAKLy8gLS0tLS0tLS0tLS0tLS0tLS0tLS0tLS0tLS0tLS0tLS0tLS0tLS0tLS0tLS0tLS0tLS0tLQoKZXhwb3J0IHR5cGUgVCA9IHsKICB4OiBzdHJpbmcsCiAgeTogbnVtYmVyLAogIHo6IGJvb2xlYW4KfQoKLy8gZXhhbXBsZXMgMSAtIGtleXMgYW5kIHByb3BlcnRpZXMKZXhwb3J0IHR5cGUgQSA9IHsgW0sgaW4ga2V5b2YgVF06IEsgfQpleHBvcnQgdHlwZSBCID0geyBbSyBpbiBrZXlvZiBUXTogVFtLXSB9CgovLyBleGFtcGxlcyAyIC0gYWRkIGFuZCByZW1vdmUgbW9kaWZpZXJzCmV4cG9ydCB0eXBlIEMgPSB7IFtLIGluIGtleW9mIFRdPzogSyB9CmV4cG9ydCB0eXBlIEQgPSB7IFtLIGluIGtleW9mIFRdPzogVFtLXSB9CmV4cG9ydCB0eXBlIEUgPSB7IFtLIGluIGtleW9mIENdLT86IENbS10gfQpleHBvcnQgdHlwZSBGID0geyBbSyBpbiBrZXlvZiBEXS0%2FOiBEW0tdIH0KCi8vIGV4YW1wbGVzIDMgLSBjb25kaXRpb25hbCBtYXAga2V5IGFuZCBwcm9wZXJ0eSB0eXBlcwpleHBvcnQgdHlwZSBHID0geyAKICBbSyBpbiBrZXlvZiBUXTogCiAgICBLIGV4dGVuZHMgJ3gnID8gJ2ZvbycgOgogICAgSyBleHRlbmRzICd5JyA%2FICdiYXInIDoKICAgIEsgZXh0ZW5kcyAneicgPyAnYmF6JyA6CiAgICBuZXZlcgp9CmV4cG9ydCB0eXBlIEggPSB7IAogIFtLIGluIGtleW9mIFRdOiAKICAgIFRbS10gZXh0ZW5kcyBzdHJpbmcgPyAnZm9vJyA6CiAgICBUW0tdIGV4dGVuZHMgbnVtYmVyID8gJ2JhcicgOgogICAgVFtLXSBleHRlbmRzIGJvb2xlYW4gPyAnYmF6JyA6CiAgICBuZXZlcgp9CgovLyBleGFtcGxlcyA0IC0gbmVzdGVkIG1hcHBpbmcKZXhwb3J0IHR5cGUgSSA9IHsKICBbWCBpbiAnYScgfCAnYiddIDogewogICAgW1kgaW4gJ2MnIHwgJ2QnXSA6IHsKICAgICAgW1ogaW4gJ2UnIHwgJ2YnXSA6IFtYLCBZLCBaXSAKICAgIH0gCiAgfSAKfQ%3D%3D)

Does any of this help?

**Comment by saxomoose** [comment]
Thank you for the reply and the clarifications.

I managed to use `Type.Index()` to address my use case:
```typescript
const AttributeReference = <T extends TSchema>(type: T, attributeName: string) => {
    const typeName = type.$id as string;
    const attributeRange = Type.Index(type, [attributeName]);

    if ("not" in attributeRange) {
        throw new Error(`The attribute ${attributeName} does not exist on the type ${typeName}.`);
    }

    return Type.Object({
        targetType: Type.Literal(typeName),
        targetAttribute: Type.Literal(attributeName),
        value: attributeRange
    });
}
```

I still feel like it would be useful to introspect the schema and obtain the `$id` of the schema used as range of a property. In the example hereunder, I would like to obtain the `$id` of the Gender schema and not the schema itself.

```typescript
const NaturalPerson = Type.Intersect([
    core.Contactable,
    Type.Object({
        givenName: Type.Array(Type.String()),
        familyName: Type.Array(Type.String()),
        gender: Gender,
    }),
], { $id: "NaturalPerson" })
``` 

If client and server use the same schemas, schema identifiers are enough to share meaning. If the schema does not have a `$id`, it is always possible to fall back on a serialization of the schema itself (e.g.: `{"type": "array", "items": {"type": "string"}}`).

I believe that the use case of passing json schema as part of the data exchange is gaining ground (cf.: [Introducing Structured Outputs in the API](https://openai.com/index/introducing-structured-outputs-in-the-api/)). Sometimes, this means sharing the full schema, but if a type registry is somehow shared, `$ids` would suffice.

**Comment by sinclairzx81** [comment]
@saxomoose Hiya,

> I managed to use Type.Index() to address my use case:

All good :)

> I still feel like it would be useful to introspect the schema and obtain the $id of the schema used as range of a property. In the example hereunder, I would like to obtain the $id of the Gender schema and not the schema itself.

I partially agree somewhat, however `$id` introspection is currently invisible to the type system (as TypeBox doesn't track the `$id` as a generic argument). The reason for this is that TypeBox interprets an `$id` is as a referential variable name (used for cross referencing types), but as TypeScript cannot derive a variable name from a type, TypeBox doesn't provide facilities for this either.

```typescript
const Foo = Type.String({ $id: 'Foo' })  // Foo is referential variable name, but isn't generic.

type Foo = string // note: there is no way to obtain `Foo` as a string literal in TypeScript
```
Keeping `$id` invisible to the type system is likely going to stay as is. However there may be updates in later revisions to make the type system aware of other options such as patternProperties, additionalProperties and if/then/else keywords.

---

Will close up this issue for now as it seems like Type.Index sounds like it resolved things :)

All the best
S
